### PR TITLE
can: isotp: Add CAN-FD support to ISO-TP

### PR DESCRIPTION
**Pull request by gramsay0**
- Add CAN-FD support to ISO-TP
- Update ISO-TP sample to work with CAN-FD
- Update existing ISO-TP unit tests to work with CAN-FD
- Add CAN-FD specific unit tests
  - Mandatory padding of frames > 8 bytes to next DLC
  - Validation that RX_DL in consecutive frames matches first frame

Thanks @dale-herman for doing most of the work

### CODE REVIEW

**Comment by alexanderwachter** [comment]
Just scrolled over it. Looks good at a first glance! You may extend the copyright notice, because it is a bigger extension of functionality.

**Comment by gramsay0** [comment]
> Just scrolled over it. Looks good at a first glance! You may extend the copyright notice, because it is a bigger extension of functionality.

@alexanderwachter thanks!

If you have time, could you please have a look over this RFC https://github.com/zephyrproject-rtos/zephyr/issues/59896 and give some feedback?
Apologies for not tagging you in it sooner, I would appreciate your input

File: tests/subsys/canbus/isotp/conformance/testcase.yaml
Comment (gramsay0) [inline tests/subsys/canbus/isotp/conformance/testcase.yaml:18]: Is there a twister filter like `dt_chosen_enabled("zephyr,canbus")` for CAN-FD?
For now, I only enabled ISO-TP CAN-FD testing on native posix

**Review by martinjaeger — CHANGES_REQUESTED** [review]
Thanks for the contribution, only got some minor comments.

May I ask you to move the changes related to `CONFIG_ISOTP_ENABLE_TX_PADDING` and `CONFIG_ISOTP_REQUIRE_RX_PADDING` into a dedicated commit? As far as I can see they are not related to CAN-FD support. This will make it easier to understand why the change was made in the future.

File: samples/subsys/canbus/isotp/Kconfig
Comment (gramsay0) [reply samples/subsys/canbus/isotp/Kconfig:21]: I think this can be simplified by using format precision to print the non-null terminated strings:

    printk("%.*s", buf->len, buf->data);

Although, I think some of the minimal string formatting libraries don't support precision.
`k_str_out` seems to be a good alternative:

    k_str_out(buf->data, buf->len);

File: subsys/canbus/isotp/isotp.c
Comment (gramsay0) [reply subsys/canbus/isotp/isotp.c:142]: My understanding is:

- `ISOTP_ENABLE_TX_PADDING` makes the device AUTOSAR conformant, by sending TX padding
- `ISOTP_REQUIRE_RX_PADDING` ensures the other devices on the bus are AUTOSAR conformant, by rejecting non-padded frames

In this case the device is transmitting a frame, so `ISOTP_ENABLE_TX_PADDING` alone would enable TX padding

I don't see a reason for one config to depend on the other

**Comment by gramsay0** [comment]
> May I ask you to move the changes related to `CONFIG_ISOTP_ENABLE_TX_PADDING` and `CONFIG_ISOTP_REQUIRE_RX_PADDING` into a dedicated commit? As far as I can see they are not related to CAN-FD support. This will make it easier to understand why the change was made in the future.

@martinjaeger Okay, will do

File: subsys/canbus/isotp/isotp.c
Comment (martinjaeger) [reply subsys/canbus/isotp/isotp.c:142]: Yes, that's correct.

But does it make sense to enforce padding for other bus participants if you don't use it yourself?

No strong opinions on that, we can also keep it as is. But the padding fix should still go into a dedicated commit as it has nothing to do with CAN-FD.

File: samples/subsys/canbus/isotp/Kconfig
Comment (martinjaeger) [reply samples/subsys/canbus/isotp/Kconfig:21]: As far as I can see we don't even need it as the terminating null character is also sent over ISO-TP (using `sizeof` and not `strlen` in the send function).

Comment (gramsay0) [reply samples/subsys/canbus/isotp/Kconfig:21]: It's the receive side that is the issue (`rx_8_0_thread()`).
The packet is returned in partial chunks and printed.
The partial chunks may not be null terminated, but the length is known

File: subsys/canbus/isotp/isotp.c
Comment (gramsay0) [reply subsys/canbus/isotp/isotp.c:142]: > But does it make sense to enforce padding for other bus participants if you don't use it yourself?

I agree that would be a strange case...
Maybe `default y if ISOTP_ENABLE_TX_PADDING` would be a good approach, as it is more of a suggestion than a hard dependency.
Yes, I will move to a separate PR

Comment (martinjaeger) [reply subsys/canbus/isotp/isotp.c:142]: That sounds like the best approach!

File: samples/subsys/canbus/isotp/Kconfig
Comment (martinjaeger) [reply samples/subsys/canbus/isotp/Kconfig:21]: Ok. I see. `k_str_out` is marked as an internal API, so I'd just go with the `printk("%.*s"...)`  if it works with the default configuration of this sample.

**Comment by gramsay0** [comment]
@martinjaeger here is the padding fix PR https://github.com/zephyrproject-rtos/zephyr/pull/60661
I'll rebase this PR once that is merged

**Review by martinjaeger — DISMISSED** [review]

File: tests/subsys/canbus/isotp/conformance/testcase.yaml
Comment (henrikbrixandersen) [reply tests/subsys/canbus/isotp/conformance/testcase.yaml]: No, you cannot filter tests based on CAN-FD support. You'll need to skip tests if runtime CAN-FD support is not available.

File: subsys/canbus/isotp/Kconfig
Comment (henrikbrixandersen) [inline subsys/canbus/isotp/Kconfig:24]: I fought hard to allow for configuring CAN-FD mode at runtime (instead of compile-time) for the CAN controllers (see https://github.com/zephyrproject-rtos/zephyr/issues/45303), so I'm a bit reluctant to add a compile-time option for ISO-TP in this regard. 

Could this be made a run-time configuration with the option to turn it all off via Kconfig (similar as to hiw `CONFIG_CAN_FD_MODE` works)?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Is ISO-TP via CAN-FD part of ISO 15765-2?

**Comment by martinjaeger** [comment]
> Is ISO-TP via CAN-FD part of ISO 15765-2?

Yes, it's part of the third edition of the spec (from 2016).

File: tests/subsys/canbus/isotp/conformance/testcase.yaml
Comment (gramsay0) [reply tests/subsys/canbus/isotp/conformance/testcase.yaml]: Good idea, I have made that change

File: subsys/canbus/isotp/Kconfig
Comment (gramsay0) [reply subsys/canbus/isotp/Kconfig:24]: Yes, that could be done, just need to determine how to best add it to the API.
Currently it would need to be passed to `isotp_bind` and the various `isotp_send` functions

@martinjaeger what are your thoughts on this?

Comment (martinjaeger) [reply subsys/canbus/isotp/Kconfig:24]: @gramsay0 Sorry for the late response.

I would suggest to add a flag to `struct isotp_msg_id` to indicate if FD frames should be used (similar to `struct can_frame` and `struct can_filter`).

The user will have to call `can_set_mode(dev, CAN_MODE_FD);` manually before using ISO-TP. I think it makes sense that the ISO-TP stack does not automatically switch to CAN-FD, as one could use the same CAN device for multiple different purposes. ISO-TP should use the mode already configured for the device.

Comment (gramsay0) [reply subsys/canbus/isotp/Kconfig:24]: @martinjaeger good idea, the `ide` flag is already in `struct isotp_msg_id` so that makes sense.
I'm on vacation for a couple weeks, feel free to make changes, otherwise I will come back to this later

Comment (martinjaeger) [reply subsys/canbus/isotp/Kconfig]: I've added a commit to make CAN-FD configurable at runtime. Please have a look @henrikbrixandersen.

**Comment by martinjaeger** [comment]
Please ignore the first two commits. The will be removed through rebasing after #61732 is merged.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
This looks good, @martinjaeger. I'd prefer if the last commit could be squashed into the previous commits to avoid the odd history, though.

File: tests/subsys/canbus/isotp/conformance/src/main.c
Comment (martinjaeger) [reply tests/subsys/canbus/isotp/conformance/src/main.c:1127]: In that case `IS_ENABLED(CONFIG_CAN_FD_MODE)` should evaluate to `0` and FD would not be enabled.

Or do you have a case in mind where we'd have one controller with FD capabilities and one without on the same board/SOC, so that we have `CONFIG_CAN_FD_MODE=y` even though one controller doesn't support it?

**Comment by martinjaeger** [comment]
> This looks good, @martinjaeger. I'd prefer if the last commit could be squashed into the previous commits to avoid the odd history, though.

I added a new commit so that the authorship of the different changes is clear. But I'm fine with squashing it if @gramsay0 agrees.

File: tests/subsys/canbus/isotp/conformance/src/main.c
Comment (henrikbrixandersen) [reply tests/subsys/canbus/isotp/conformance/src/main.c:1127]: `CONFIG_CAN_FD_MODE` can be enabled whether the CAN controller supports CAN-FD or not.

Comment (martinjaeger) [reply tests/subsys/canbus/isotp/conformance/src/main.c:1127]: Yeah, that's exactly how the test is run at the moment, e.g. with `native_posix`. The first case is with default setup `CONFIG_CAN_FD_MODE=n` (even though it would be supported) and the two other cases with extra config `CONFIG_CAN_FD_MODE=y` and some different max data length settings.

Or maybe I misunderstand your question/request?

Comment (henrikbrixandersen) [reply tests/subsys/canbus/isotp/conformance/src/main.c:1127]: The tests need to be able to run on any CAN controller, regardless of `CONFIG_CAN_FD_MODE` being `y` or `n` and whether the CAN controller driver supports CAN-FD or not.

I'd suggest using `can_get_capabilities()` to determine if `CAN_MODE_FD` should be passed to `can_set_mode()` . That does not require any `#ifdef` on `CONFIG_CAN_FD_MODE`.

File: subsys/canbus/isotp/isotp.c
Comment (dale-herman) [inline subsys/canbus/isotp/isotp.c]: Should brs (data bitrate switch) be an independent setting dependent on CAN-FD being enabled? This requires the databit rate baud and sampling to be set.

Comment (henrikbrixandersen) [reply subsys/canbus/isotp/isotp.c]: Good idea. Perhaps we would be better off using a bitmask of options?

Comment (martinjaeger) [reply subsys/canbus/isotp/isotp.c]: Good point. I've added the `.brs` as as another parameter to the `struct isotp_msg_id`.

I decided against using the existing flags like `CAN_FRAME_FDF` and `CAN_FRAME_BRS`, as we need to apply the configuration both to the filter and the frames. The CAN driver uses different flags for frame and filter, so it would be a bit confusing.

File: tests/subsys/canbus/isotp/conformance/src/main.c
Comment (martinjaeger) [reply tests/subsys/canbus/isotp/conformance/src/main.c:1146]: It was slightly more complicated than just checking with `can_get_capabilities` and applying the mode accordingly. Now we have the existing `main` test suite, which is skipped if FD is enabled for a controller that does not have FD capabilities. To check that correct errors are returned in such cases, I added another test suite `canfd_notsup.c`.

File: subsys/canbus/isotp/isotp.c
Comment (martinjaeger) [reply subsys/canbus/isotp/isotp.c]: Ok, that makes sense and I double-checked if we could introduce the flags as part of this PR. However, it would require changes to code unrelated to CAN-FD, as we are currently using two other parameters for the ISO-TP extended addressing and fixed addressing features.

For this reason, I suggest to switch to flags in a follow-up PR.

File: tests/subsys/canbus/isotp/conformance/src/canfd_notsup.c
Comment (henrikbrixandersen) [inline tests/subsys/canbus/isotp/conformance/src/canfd_notsup.c:1]: This logic is reversed to me. We should aim for a generic ISO-TP conformance test suite that is always run, and then have an ISO-TP via CAN-FD test suite that is run if both `CONFIG_CAN_FD_MODE` is enabled _and_ the `zephyr,chosen` CAN controller capabilities includes CAN-FD support, otherwise is should be skipped (or even not compiled-in in the case of `CONFIG_CAN_FD_MODE=n`).

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: subsys/canbus/isotp/Kconfig
Comment (martinjaeger) [reply subsys/canbus/isotp/Kconfig:29]: Yes, it is part of the standard:

image

File: subsys/canbus/isotp/isotp.c
Comment (martinjaeger) [reply subsys/canbus/isotp/isotp.c]: Change of mind. Let's get the flags in first: #62255

I'll get back to this PR afterwards.

**Comment by gramsay0** [comment]
> > This looks good, @martinjaeger. I'd prefer if the last commit could be squashed into the previous commits to avoid the odd history, though.
> 
> I added a new commit so that the authorship of the different changes is clear. But I'm fine with squashing it if @gramsay0 agrees.

Yep, I'm ok with that.
BTW thanks for carrying this on 👍

File: subsys/canbus/isotp/isotp.c
Comment (alexanderwachter) [inline subsys/canbus/isotp/isotp.c:45]: Not sure if this is going to be efficient. In C++, we can rely on return value optimization (copy elision), but not in C.
I'm not sure if the inline forces to not copy the struct on return.

File: include/zephyr/canbus/isotp.h
Comment (gramsay0) [inline include/zephyr/canbus/isotp.h:186]: I was thinking that doing this at runtime could be useful 👍
Could send lower priority messages as smaller frames to reduce high priority latency.

Maybe also interpret 0 as 64 if CAN-FD is enabled, as that is probably the default expected behavior

File: include/zephyr/canbus/isotp.h
Comment (gramsay0) [inline include/zephyr/canbus/isotp.h:159]: Typo: CAN_FRAME_FDF -> ISOTP_MSG_FDF

File: subsys/canbus/isotp/isotp.c
Comment (gramsay0) [inline subsys/canbus/isotp/isotp.c:1257]: This switch should have a `default:` case that returns an error

File: tests/subsys/canbus/isotp/conformance/src/main.c
Comment (gramsay0) [inline tests/subsys/canbus/isotp/conformance/src/main.c:374]: Not a bug from these changes but `&des_frames[i]` here should be `frames`, and the same thing a few lines below this.
It has no impact on the existing tests, but maybe worth tidying up

**Review by gramsay0 — COMMENTED** [review]
Good work!

File: subsys/canbus/isotp/isotp.c
Comment (martinjaeger) [reply subsys/canbus/isotp/isotp.c]: I'm absolutely not fluent in assembly. I can at least see that it's not inlined, but kept as a separate function call. Not sure if the function copies the struct on return or not.

<details>
<summary>Disassembly for arm-none-eabi</summary>

```asm
08008008 <prepare_frame>:
{
 8008008:	b538      	push	{r3, r4, r5, lr}
 800800a:	460d      	mov	r5, r1
	struct can_frame frame = {
 800800c:	220d      	movs	r2, #13
 800800e:	2100      	movs	r1, #0
{
 8008010:	4604      	mov	r4, r0
	struct can_frame frame = {
 8008012:	3003      	adds	r0, #3
 8008014:	f001 fb65 	bl	80096e2 <memset>
		.flags = (addr->flags & ISOTP_MSG_IDE) != 0 ? CAN_FRAME_IDE : 0,
 8008018:	79a9      	ldrb	r1, [r5, #6]
		.id = addr->ext_id
 800801a:	682b      	ldr	r3, [r5, #0]
	if ((addr->flags & ISOTP_MSG_FDF) != 0) {
 800801c:	0708      	lsls	r0, r1, #28
		.id = addr->ext_id
 800801e:	f3c3 021c 	ubfx	r2, r3, #0, #29
		.flags = (addr->flags & ISOTP_MSG_IDE) != 0 ? CAN_FRAME_IDE : 0,
 8008022:	f3c1 0380 	ubfx	r3, r1, #2, #1
	if ((addr->flags & ISOTP_MSG_FDF) != 0) {
 8008026:	d505      	bpl.n	8008034 <prepare_frame+0x2c>
		if ((addr->flags & ISOTP_MSG_BRS) != 0) {
 8008028:	06c9      	lsls	r1, r1, #27
		frame.flags |= CAN_FRAME_FDF;
 800802a:	bf54      	ite	pl
 800802c:	f043 0304 	orrpl.w	r3, r3, #4
			frame.flags |= CAN_FRAME_BRS;
 8008030:	f043 030c 	orrmi.w	r3, r3, #12
	return frame;
 8008034:	6821      	ldr	r1, [r4, #0]
 8008036:	7163      	strb	r3, [r4, #5]
 8008038:	f362 011c 	bfi	r1, r2, #0, #29
 800803c:	6021      	str	r1, [r4, #0]
}
 800803e:	4620      	mov	r0, r4
 8008040:	bd38      	pop	{r3, r4, r5, pc}
```
</details>

Even if it takes a copy, it's also a matter of code clarity and reduced duplication. So I'd probably keep it as a function call.

File: include/zephyr/canbus/isotp.h
Comment (alexanderwachter) [reply include/zephyr/canbus/isotp.h:186]: True.
As far as I remember, 8 is mandatory for classic CAN.

Comment (martinjaeger) [reply include/zephyr/canbus/isotp.h:186]: Yes, I know 8 is mandatory for classical CAN, that's why values < 8 are not in that list. I would rather not ignore it for classical CAN, as you don't want people to choose 6 (or whatever) and implicitly get 8. Default 64 for CAN-FD will come with the next rebase once all the CI issues on main are fixed.

File: tests/subsys/canbus/isotp/conformance/src/main.c
Comment (martinjaeger) [reply tests/subsys/canbus/isotp/conformance/src/main.c]: Moved back to setup.

File: subsys/canbus/isotp/isotp.c
Comment (martinjaeger) [reply subsys/canbus/isotp/isotp.c]: Done. It's now a void function with pointer argument.

File: include/zephyr/canbus/isotp.h
Comment (martinjaeger) [reply include/zephyr/canbus/isotp.h]: Good spot. Fixed.

File: tests/subsys/canbus/isotp/conformance/src/main.c
Comment (martinjaeger) [reply tests/subsys/canbus/isotp/conformance/src/main.c]: Yeah, you are right. I've added a commit to fix this bug.

File: tests/subsys/canbus/isotp/conformance/src/main.c
Comment (martinjaeger) [reply tests/subsys/canbus/isotp/conformance/src/main.c]: This is now considered using a test-specific `CONFIG_TEST_USE_CAN_FD_MODE`. `CONFIG_CAN_FD_MODE` can be switched on or off independent of that.

File: tests/subsys/canbus/isotp/conformance/src/canfd_notsup.c
Comment (martinjaeger) [reply tests/subsys/canbus/isotp/conformance/src/canfd_notsup.c:1]: I agree the naming was not ideal. It is now called `mode_check.c` and it is run for any configuration. The main test suite is also run for any configuration, but skipped if it is configured to use CAN-FD even though it's not supported by the controller.

Note that it is still run if `CONFIG_CAN_FD_MODE=y`, but `CONFIG_TEST_USE_CAN_FD_MODE=n`.

I chose this approach instead of a dedicated CAN-FD test suite because most of the ISO-TP functionality stays the same, CAN-FD only introduces subtile changes in some locations. If we wanted to have one complete test suit for FD and one for classical CAN, we'd end up with lots of code duplication.

**Comment by martinjaeger** [comment]
I hope that all review comments are addressed now and unrelated CI failures are gone.

The ISO-TP test suites and sample were successfully tested on the following boards:
- `native_posix`
- `nucleo_g474re` (CAN-FD support)
- `nucleo_f072rb` (classical CAN only)

File: tests/subsys/canbus/isotp/conformance/testcase.yaml
Comment (henrikbrixandersen) [inline tests/subsys/canbus/isotp/conformance/testcase.yaml:9]: These cases significantly increase the time required for running the ISO-TP conformance test suite on real hardware. Can the tests be optimized in time somehow?

Example of time spent running this test suite on five random boards on my desk:
```
INFO    -  6/35 frdm_k64f                 tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_64 PASSED (device: 0240000047784e45004e9003d9170022e561000097969900, 8.151s)
INFO    -  7/35 lpcxpresso55s36           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_64 PASSED (device: 001069887263, 14.602s)
INFO    -  8/35 stm32f3_disco             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_64 PASSED (device: 066EFF555052836687113518, 4.123s)
INFO    -  9/35 mimxrt1060_evkb           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_64 PASSED (device: 0229000033f4cd1100000000000000000000000097969905, 33.479s)
INFO    - 10/35 nucleo_g474re             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_64 PASSED (device: 0041001B3438510B34313939, 13.539s)
INFO    - 11/35 frdm_k64f                 tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_32 PASSED (device: 0240000047784e45004e9003d9170022e561000097969900, 10.311s)
INFO    - 12/35 lpcxpresso55s36           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_32 PASSED (device: 001069887263, 11.119s)
INFO    - 13/35 stm32f3_disco             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_32 PASSED (device: 066EFF555052836687113518, 3.966s)
INFO    - 14/35 mimxrt1060_evkb           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_32 PASSED (device: 0229000033f4cd1100000000000000000000000097969905, 37.112s)
INFO    - 15/35 nucleo_g474re             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_32 PASSED (device: 0041001B3438510B34313939, 13.679s)
INFO    - 16/35 frdm_k64f                 tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.unused PASSED (device: 0240000047784e45004e9003d9170022e561000097969900, 16.917s)
INFO    - 17/35 lpcxpresso55s36           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.unused PASSED (device: 001069887263, 10.552s)
INFO    - 18/35 stm32f3_disco             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.unused PASSED (device: 066EFF555052836687113518, 14.584s)
INFO    - 19/35 mimxrt1060_evkb           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.unused PASSED (device: 0229000033f4cd1100000000000000000000000097969905, 33.130s)
INFO    - 20/35 nucleo_g474re             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.unused PASSED (device: 0041001B3438510B34313939, 14.217s)
INFO    - 21/35 frdm_k64f                 tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance PASSED (device: 0240000047784e45004e9003d9170022e561000097969900, 17.851s)
INFO    - 22/35 lpcxpresso55s36           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance PASSED (device: 001069887263, 11.913s)
INFO    - 23/35 mimxrt1060_evkb           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance PASSED (device: 0229000033f4cd1100000000000000000000000097969905, 31.035s)
INFO    - 24/35 stm32f3_disco             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance PASSED (device: 066EFF555052836687113518, 14.670s)
INFO    - 25/35 nucleo_g474re             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance PASSED (device: 0041001B3438510B34313939, 14.054s)
```

Do we really need to test with different TX_DLs on all boards or could a simple unit-like test (e.g. on POSIX) be sufficient for the different configurations? I mean, these tests validate the protocol implementation, not as much as the protocol interaction with the particular CAN controller?

Another option could be to tag some of these tests as "slow".

**Review by henrikbrixandersen — DISMISSED** [review]
Looks good. A couple of minor observations:

**Review by dale-herman — APPROVED** [review]
Looks good.

**Review by alexanderwachter — DISMISSED** [review]

File: tests/subsys/canbus/isotp/conformance/src/mode_check.c
Comment (martinjaeger) [reply tests/subsys/canbus/isotp/conformance/src/mode_check.c]: Fixed.

File: tests/subsys/canbus/isotp/conformance/testcase.yaml
Comment (martinjaeger) [reply tests/subsys/canbus/isotp/conformance/testcase.yaml:9]: There will be some more upcoming ISO-TP updates. Let's leave the optimization of test time for a future PR.

**Comment by martinjaeger** [comment]
Fixed the typo in the test naming identified by @henrikbrixandersen. Please re-approve once again, @alexanderwachter . Thanks!

**Review by martinjaeger — DISMISSED** [review]
Approval and thanks for the parts of the work @gramsay0 and @dale-herman have done. Also thanks for all reviews.

**Review by henrikbrixandersen — APPROVED** [review]

**Review by alexanderwachter — APPROVED** [review]
Thanks Martin!

**Pull request merged**
